packages/cli/src/events/WorkflowStatistics.ts

```diff
@@ -0,0 +1,100 @@
+import { INode, IRun, IWorkflowBase } from 'n8n-workflow';
+import { Db, InternalHooksManager } from '..';
+import { StatisticsNames } from '../databases/entities/WorkflowStatistics';
+import { getWorkflowOwner } from '../UserManagement/UserManagementHelper';
+
+export async function workflowExecutionCompleted(
+	workflowData: IWorkflowBase,
+	runData: IRun,
+): Promise<void> {
+	// Determine the name of the statistic
+	const finished = runData.finished ? runData.finished : false;
+	const manual = runData.mode === 'manual';
+	let name: StatisticsNames;
+
+	if (finished) {
+		if (manual) name = StatisticsNames.manualSuccess;
+		else name = StatisticsNames.productionSuccess;
+	} else {
+		if (manual) name = StatisticsNames.manualError;
+		else name = StatisticsNames.productionError;
+	}
+
+	// Get the workflow id
+	let workflowId: number;
+	try {
+		workflowId = parseInt(workflowData.id as string, 10);
+		if (isNaN(workflowId)) throw new Error('not a number');
+	} catch (error) {
+		console.error(`Error "${error as string}" when casting workflow ID to a number`);
+		return;
+	}
+
+	// Try insertion and if it fails due to key conflicts then update the existing entry instead
+	try {
```

Comment:
Could we use .upsert() here instead of waiting for an exception? (see https://typeorm.io/repository-api)

Reply:
Upsert wasn't working when we tried it because the value that would be updated would be a count field that wouldn't initially exist for insert, since we have to start it at 1. It was something we did try before but I wasn't able to get it working

Reply:
I think that was the issue with onConflict in queryBuilder, but upsert ended up being unusable since you can't use `set count = count + 1` for both insert and update which is why they have to be separate